src/nnet3/nnet-chain-training.cc

```diff
@@ -68,23 +69,81 @@ void NnetChainTrainer::Train(const NnetChainExample &chain_eg) {
                              &request);
   const NnetComputation *computation = compiler_.Compile(request);
 
-  NnetComputer computer(nnet_config.compute_config, *computation,
+  if (nnet_config.backstitch_training_scale > 0.0 && num_minibatches_processed_
+      % nnet_config.backstitch_training_interval == 0) {
+    // backstitch training is incompatible with momentum > 0
+    KALDI_ASSERT(nnet_config.momentum == 0.0);
+    FreezeNaturalGradient(true, delta_nnet_);
+    bool is_backstitch_step = true;
+    srand(srand_seed_ + num_minibatches_processed_);
+    ResetGenerators(nnet_);
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+    FreezeNaturalGradient(false, delta_nnet_); // un-freeze natural gradient
+    is_backstitch_step = false;
+    srand(srand_seed_ + num_minibatches_processed_);
+    ResetGenerators(nnet_);
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+  } else {
+    // conventional training
+    bool is_backstitch_step = false;
+    TrainInternal(chain_eg, *computation, is_backstitch_step);
+  }
+
+  num_minibatches_processed_++;
+}
+
+void NnetChainTrainer::TrainInternal(const NnetChainExample &eg,
```

Comment:
I think it would make the code easier to understand if you split TrainInternal into two functions:
TrainInternal(), with an interface like the current function but with no is_backstitch_step argument, and
TrainInternalBackstitch(),  taking the is_backstitch_step1 argument.  Currently the code relies on the condition
```
if (nnet_config.backstitch_training_scale > 0.0 && num_minibatches_processed_
      % nnet_config.backstitch_training_interval == 0) {
```
being identical in 2 places, which is a bit ugly.  Obviously simplify in reasonable ways, e.g. momentum
is not allowed in backstitch, so no need to refer to it at all in the backstitch version of the function.